orderer/common/filerepo/filerepo.go

```diff
@@ -0,0 +1,169 @@
+	}
+
+	files, err := ioutil.ReadDir(fileRepoDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Remove existing transient files in the repo
+	for _, f := range files {
+		transientFilePattern := "*" + fileSuffix + "~"
+		isTransientFile, err := filepath.Match(transientFilePattern, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if isTransientFile {
+			os.Remove(filepath.Join(fileRepoDir, f.Name()))
+		}
+	}
+
+	if err := fileutil.SyncDir(fileRepoDir); err != nil {
+		return nil, err
+	}
+
+	return &Repo{
+		transientFileMarker: "~",
+		fileSuffix:          fileSuffix,
+		fileRepoDir:         fileRepoDir,
+	}, nil
+}
+
+// Save atomically persists the content to suffix/baseName+suffix file by first writing it
+// to a tmp file marked by the transientFileMarker and then moves the file to the final
+// destination indicated by the FileSuffix.
+func (r *Repo) Save(baseName string, content []byte) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	fileName := fmt.Sprintf("%s.%s", baseName, r.fileSuffix)
+	dest := filepath.Join(r.fileRepoDir, fileName)
+
+	if _, err := os.Stat(dest); err == nil {
+		return errors.Errorf("file already exists at %s", dest)
+	}
+
+	tmpFileName := fileName + r.transientFileMarker
+	if err := fileutil.CreateAndSyncFileAtomically(r.fileRepoDir, tmpFileName, fileName, content, repoFilePermPrivateRW); err != nil {
+		return err
+	}
+
+	return fileutil.SyncDir(r.fileRepoDir)
+}
+
+// Remove removes the file associated with baseName from the file system.
+func (r *Repo) Remove(baseName string) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	fileName := fmt.Sprintf("%s.%s", baseName, r.fileSuffix)
+	filePath := filepath.Join(r.fileRepoDir, fileName)
+
+	if err := os.RemoveAll(filePath); err != nil {
+		return err
+	}
+
+	return fileutil.SyncDir(r.fileRepoDir)
+}
+
+// Read reads the file in the fileRepo associated with baseName's contents.
+func (r *Repo) Read(baseName string) ([]byte, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	fileName := fmt.Sprintf("%s.%s", baseName, r.fileSuffix)
+	filePath := filepath.Join(r.fileRepoDir, fileName)
+	return ioutil.ReadFile(filePath)
+}
+
+// List parses the directory and produce a list of file names, filtered by suffix.
+func (r *Repo) List() ([]string, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var repoFiles []string
+
+	files, err := ioutil.ReadDir(r.fileRepoDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, f := range files {
+		isFileSuffix, err := filepath.Match("*"+r.fileSuffix, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if isFileSuffix {
+			repoFiles = append(repoFiles, f.Name())
+		}
+	}
+
+	return repoFiles, nil
+}
+
+// FileToBaseName strips the suffix from the file name to get the associated channel name.
+func (r *Repo) FileToBaseName(fileName string) string {
+	baseFile := filepath.Base(fileName)
+
+	return strings.TrimSuffix(baseFile, "."+r.fileSuffix)
+}
+
+func validateFileSuffix(fileSuffix string) error {
+	if len(fileSuffix) <= 0 {
```

Comment:
Nit. as it can never be < 0
```suggestion
	if len(fileSuffix) == 0 {
```